db/hints/internal/hint_storage.cc

```diff
@@ -50,16 +50,24 @@ using hints_ep_segments_map = std::unordered_map<unsigned, std::list<fs::path>>;
 // map: IP -> map: shard -> segments
 using hints_segments_map = std::unordered_map<sstring, hints_ep_segments_map>;
 
-future<> scan_for_hints_dirs(const fs::path& hints_directory, std::function<future<> (fs::path dir, directory_entry de, unsigned shard_id)> f) {
-    return lister::scan_dir(hints_directory, lister::dir_entry_types::of<directory_entry_type::directory>(), [f = std::move(f)] (fs::path dir, directory_entry de) mutable {
+/// The caller must ensure that the passed functor's captures
+/// are kept alive at least until the returned future resolves.
+/// It also applies to the functor itself if it's passed as an lvalue reference.
+template <typename Func>
+    requires std::invocable<Func, fs::path, directory_entry, unsigned>
+future<> scan_for_hints_dirs(const fs::path& hint_directory, Func&& func) {
+    return lister::scan_dir(hint_directory, lister::dir_entry_types::of<directory_entry_type::directory>(),
+            [func = std::forward<Func>(func)] (fs::path dir, directory_entry de) mutable {
         unsigned shard_id;
+
         try {
-            shard_id = std::stoi(de.name.c_str());
+            shard_id = std::stoi(de.name);
         } catch (std::invalid_argument& ex) {
             manager_logger.debug("Ignore invalid directory {}", de.name);
             return make_ready_future<>();
         }
-        return f(std::move(dir), std::move(de), shard_id);
+
+        return futurize_invoke(std::forward<Func>(func), std::move(dir), std::move(de), shard_id);
```

Comment:
Is `futurize_invoke` here necessary?

Reply:
Yes. `lister::scan_dir` requires that the passed functor return `future<>`, but it might make sense to pass a void function to the function too. `futurize_invoke` is a way to ensure those functors are also valid arguments.